File: src/api/assignments.ts
import { type SessionHandle, type Assignment, TabLocation } from "~/models";
import { decodeAssignment } from "~/decoders/assignment";
import { homeworkFromIntervals, homeworkFromWeek } from "./private/homework";

const decoder = (session: SessionHandle, data: any): Array<Assignment> => {
  return data.ListeTravauxAFaire.V.map((homework: any) => decodeAssignment(homework, session));
};

export const assignmentsFromWeek = async (session: SessionHandle, weekNumber: number, extendsToWeekNumber?: number): Promise<Array<Assignment>> => {
  const reply = await homeworkFromWeek(session, TabLocation.Assignments, weekNumber, extendsToWeekNumber);
  return decoder(session, reply);
};

export const assignmentsFromIntervals = async (session: SessionHandle, startDate: Date, endDate: Date): Promise<Array<Assignment>> => {
  const reply = await homeworkFromIntervals(session, TabLocation.Assignments, startDate, endDate);
  // Only keep items assignments are in the intervals.
  return decoder(session, reply).filter((homework) => startDate <= homework.deadline && homework.deadline <= endDate);
};

File: library/src/commonMain/kotlin/api/assignments.kt
package api

import api.private.homeworkFromIntervals
import api.private.homeworkFromWeek
import decoders.decodeAssignment
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import models.Assignment
import models.SessionHandle
import models.SessionInformation
import models.TabLocation

val decoder: (sessionInformation: SessionInformation, data: JsonObject) -> List<Assignment> = {sessionInformation, data ->
    data["ListeTravauxAFaire"]!!.jsonObject["V"]!!.jsonArray.map { decodeAssignment(it.jsonObject, sessionInformation) }
}

suspend fun assignmentsFromWeek (sessionInformation: SessionInformation, weekNumber: Int, extendsToWeekNumber: Int? = null): List<Assignment> {
    val reply = homeworkFromWeek(sessionInformation, TabLocation.Assignments, weekNumber, extendsToWeekNumber)
    return decoder(sessionInformation, reply)
}

suspend fun assignmentsFromIntervals (session: SessionHandle, startDate: LocalDateTime, endDate: LocalDateTime): List<Assignment> {
    val reply = homeworkFromIntervals(session, TabLocation.Assignments, startDate, endDate)
    // Only keep items assignments are in the intervals.
    return decoder(session.information, reply).filter { it.deadline in startDate..endDate }
}